src/noit_message_decoder.c

```diff
@@ -174,32 +219,356 @@ int noit_message_decoder_parse_line(const char *payload, int payload_len,
+  return squash_index_left + 1;
+}
+
+ssize_t
+noit_metric_tags_parse(const char *tagnm, size_t tagnmlen,
+                       noit_metric_tag_t *tags, size_t tag_count,
+                       size_t *canonical_size_out) {
+  if(tagnmlen == 0) return 0;
+
+  size_t rval = 0;
+  while(tagnmlen > 0 && rval < tag_count) {
+    const char *next = noit_metric_tags_parse_one(tagnm, tagnmlen, &tags[rval]);
+    if(!next) return -1;
+
+    tagnmlen -= (next - tagnm);
+    tagnm = next;
+    rval++;
+    if(tagnmlen > 0) {
+      /* there's string left, we require it to be ",<next_tag>". */
+      if(*tagnm != ',' || tagnmlen == 1) return -1;
+      tagnm++;
+      tagnmlen--;
+    }
+  }
+
+  return (ssize_t) noit_metric_tags_compact(tags, rval, canonical_size_out);
+}
+
+ssize_t
+noit_metric_tags_canonical(const noit_metric_tag_t *tags,
+                           size_t tag_count, char *tagnm, size_t tagnmlen) {
+  if(!tag_count) return 0;
+  size_t rval = 0;
+  while(tag_count > 0) {
+    if(tagnmlen < tags[0].total_size) return -1;
+    memcpy(tagnm, tags[0].tag, tags[0].total_size);
+    tagnm += tags[0].total_size;
+    tagnmlen -= tags[0].total_size;
+    rval += tags[0].total_size;
+    tags++;
+    tag_count--;
+    if(tag_count > 0) {
+      if (tagnmlen <= 0) return -1;
+      *tagnm++ = ',';
+      rval++;
+      tagnmlen--;
+    }
+  }
+  return rval;
+}
+
+char *
+noit_metric_tagset_from_tags(noit_metric_tagset_t *lookup,
+                             noit_metric_tag_t *tags,
+                             size_t tag_count,
+                             size_t canonical_size) {
+  /* the canonical representation does not include a trailing NULL, but we make room for one
+   * anyway... helps our display logic. */
+  char *canonical = malloc(canonical_size + 1);
+  if(!canonical) return NULL;
+  if(noit_metric_tags_canonical(tags, tag_count, canonical, canonical_size) != canonical_size) {
+    free((void *) canonical);
+    return NULL;
+  }
+  canonical[canonical_size] = '\0';
+  /* reparse the tags, so that they point into the canonical string.
+   * (this simplifies life-time management for the lookup_t
+   * structure.) */
+  noit_metric_tags_parse(canonical, canonical_size, tags, tag_count, &canonical_size);
+
+  return canonical;
+}
+
+int
+noit_metric_tagset_init(noit_metric_tagset_t *lookup,
+                        const char *tagnm, size_t tagnmlen) {
+  memset((void *) lookup, '\0', sizeof(*lookup));
+  size_t tag_count = noit_metric_tags_count(tagnm, tagnmlen);
+  if(tag_count == 0) {
+    lookup->tags = 0;
+    lookup->tag_count = 0;
+    lookup->canonical_size = 0;
+    return 0;
+  }
+  noit_metric_tag_t *tags = (noit_metric_tag_t *) calloc(tag_count, sizeof(noit_metric_tag_t));
+  size_t canonical_size;
+  ssize_t parsed_tag_count = noit_metric_tags_parse(tagnm, tagnmlen, tags, tag_count, &canonical_size);
+  if(parsed_tag_count <= 0) {
+    free((void *) tags);
+    return -1;
+  }
+  lookup->tags = tags;
```

Comment:
suggest `lookup->tags = realloc(tags), tag_count * sizeof(noit_metric_tag_t *)` here. number of elements in `tags` array might have shrunk since initial allocation, and I believe `realloc` is fast when reducing size.